resources/translations.ts

```diff
@@ -271,3 +272,35 @@ export const translateText = (
   replaceLang: Lang,
   replacements?: TimelineReplacement[],
 ): string => translateWithReplacements(text, 'replaceText', replaceLang, replacements).text;
+
+// Translates a timeline or trigger regex for a given language.
+export const translateRegexBuildParam = <T extends NetParams[keyof NetProps]>(
+  param: T,
+  replaceLang: Lang,
+  replacements?: TimelineReplacement[],
+): T => {
+  if ('source' in param)
```

Comment:
Could you export and re-use `keysThatRequiresTranslation` from `netregexes.ts`?

Reply:
Oh, didn't find this array before.

Reply:
I'd like to use this array here, but I can't pass type check here 😕

Reply:
Try to do something like this:
```ts
// Translates a timeline or trigger regex for a given language.
export const translateRegexBuildParam = <T extends NetParams[keyof NetProps]>(
  param: T,
  replaceLang: Lang,
  replacements?: TimelineReplacement[],
): T => {
  for (const key of (keysThatRequireTranslation as KeysThatRequireTranslation[])) {
    if (key in param)
      param[key] = translateField(param[key], replaceLang, replacements);
  }

  return param;
};
```

```text
resources/translations.ts:284:7 - error TS7053: Element implicitly has an 'any' type because expression of type 'KeysThatRequireTranslation' can't be used to index type 'Partial<Record<"type" | "code" | "name" | "line", string | string[]> & { timestamp: string; capture: boolean; }> | Partial<Record<"type" | "name" | "id", string | string[]> & { ...; }> | ... 28 more ... | Partial<...>'.
  Property 'name' does not exist on type 'Partial<Record<"type" | "code" | "name" | "line", string | string[]> & { timestamp: string; capture: boolean; }> | Partial<Record<"type" | "name" | "id", string | string[]> & { ...; }> | ... 28 more ... | Partial<...>'.

284       param[key] = translateField(param[key], replaceLang, replacements);
          ~~~~~~~~~~

resources/translations.ts:284:35 - error TS7053: Element implicitly has an 'any' type because expression of type 'KeysThatRequireTranslation' can't be used to index type 'Partial<Record<"type" | "code" | "name" | "line", string | string[]> & { timestamp: string; capture: boolean; }> | Partial<Record<"type" | "name" | "id", string | string[]> & { ...; }> | ... 28 more ... | Partial<...>'.
  Property 'name' does not exist on type 'Partial<Record<"type" | "code" | "name" | "line", string | string[]> & { timestamp: string; capture: boolean; }> | Partial<Record<"type" | "name" | "id", string | string[]> & { ...; }> | ... 28 more ... | Partial<...>'.

284       param[key] = translateField(param[key], replaceLang, replacements);
```

any help？

Reply:
never mind, works fine now.

Reply:
What about:

```typescript
// Translates a timeline or trigger regex for a given language.
export const translateRegexBuildParam = <T extends TriggerTypes>(
  param: NetParams[T] & Partial<Record<KeysThatRequireTranslation, string | string[]>>,
  replaceLang: Lang,
  replacements?: TimelineReplacement[],
): NetParams[T] => {
  for (const key of keysThatRequireTranslation) {
    if (key in param)
      param[key] = translateField(param[key], replaceLang, replacements);
  }

  return param;
};
```

Also:
```typescript
export const keysThatRequireTranslation = [
  'ability',
  'name',
  'source',
  'target',
  'line',
] as const;
export type KeysThatRequireTranslation = typeof keysThatRequireTranslation[number];
```

Reply:
No, you can't.

In this case `keysThatRequireTranslation` is now a `ReadonlyArray<KeysThatRequireTranslation[]>` and you can't do `keysThatRequireTranslation.includes(string)` anymore

Reply:
Here I think is a issue of typescript itself. An `ReadonlyArray<T>` has a `includes(item: T): boolean` method, it's not a type narrow function `includes(item: string): item is T`

Reply:
> No, you can't.
> 
> In this case `keysThatRequireTranslation` is now a `ReadonlyArray<KeysThatRequireTranslation[]>` and you can't do `keysThatRequireTranslation.includes(string)` anymore

```typescript
const keysThatRequireTranslationAsConst = [
  'ability',
  'name',
  'source',
  'target',
  'line',
] as const;
export const keysThatRequireTranslation: readonly string[] = keysThatRequireTranslationAsConst;
export type KeysThatRequireTranslation = typeof keysThatRequireTranslationAsConst[number];
```

This will work and there's no need for a "keep these in sync" comment.